packages/math/src/pow.rs

```diff
@@ -0,0 +1,202 @@
+use cosmwasm_std::{Decimal256, Uint256};
+
+use crate::error::MathError;
+
+const DEC_PRECISION: u128 = 1_000_000_000_000_000_000u128;
+const POW_PRECISION: u128 = 10_000_000_000u128;
+
+/// Calculates the power of a base to an exponent
+/// This function contains Osmosis's `pow` implementation as closely as possible
+pub fn pow(base: Decimal256, exp: Decimal256) -> Result<Decimal256, MathError> {
+    if base.is_zero() {
+        return Err(MathError::NegativeBase);
+    }
+
+    if base >= Decimal256::from_ratio(2u64, 1u64) {
+        return Err(MathError::BaseTooLarge);
+    }
+
+    if exp == Decimal256::one() {
+        return Ok(base);
+    }
+
+    let integer = exp.to_uint_floor();
+    let fractional = {
+        let x = exp.atomics();
+        let y = Uint256::from(DEC_PRECISION);
+
+        Decimal256::from_atomics(x - (y * (x / y)), base.decimal_places())
+    }?;
+
+    let integer_pow = pow_int(base, integer);
+    if fractional.is_zero() {
+        return Ok(integer_pow);
+    }
+
+    let fractional_pow = pow_approx(
+        base,
+        fractional,
+        Decimal256::from_atomics(POW_PRECISION, base.decimal_places())?,
+    )?;
+
+    Ok(integer_pow * fractional_pow)
+}
+
+fn pow_int(mut base: Decimal256, power: Uint256) -> Decimal256 {
+    if power <= Uint256::one() {
+        return Decimal256::one();
+    }
+
+    let mut tmp = Decimal256::one();
+
+    let mut i = power;
+
+    loop {
+        if i <= Uint256::one() {
+            break;
+        }
+
+        if i % Uint256::from(2u64) != Uint256::zero() {
+            tmp *= base;
+        }
+
+        i /= Uint256::from(2u64);
+
+        base *= base;
+    }
+
+    base * tmp
+}
+
+fn pow_approx(
```

Comment:
👏👏👏👍

Reply:
We can apply this approx logic in simulating stable swap pool :+1: